### ISSUE TITLE
Validate that our codegen won't be broken in CoreCLR RTM

### ISSUE THREAD

**Issue by sergeybykov** [dotnet/orleans#1741]
`FormatterServices.GetUninitializedObject()` was made internal in CoreCLR RC1, which broke our codegen. There is an issue https://github.com/dotnet/corefx/issues/7938 to bring it back. This is supposed to happen in RTM, but not in RC2.

The workaround that was suggested in the meantime is to invoke `FormatterServices.GetUninitializedObject()` via reflection as is done in https://github.com/dotnet/corefx/blob/0444706f7ab72e4cdb649af6d21de1c2f6d334f9/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs#L33.

If we tweak our codegen accordingly, we should be able to test if the rest of our codegen will compile fine against RC2. If we discover other issues, we'll have a time window, albeit a short one, to bring them to the CoreCLR team, so that they have a chance to address such issues before CoreCLR RTM.

**Comment by dVakulen** [comment]
Regarding 'FormatterServices.GetUninitializedObject()' - I've already implemented workaround in the https://github.com/dotnet/orleans/pull/1122.

**Comment by sergeybykov** [comment]
@dVakulen But that's a different workaround, that uses `Activator.CreateInstance()` instead of `FormatterServices.GetUninitializedObject()`, right? It requires a default constructor to be public?

**Comment by dVakulen** [comment]
@sergeybykov No, it requires constructor that doesn't check if it's parameters are nulls, as it passes `default(T)` for each ctor parameter.

**Comment by sergeybykov** [comment]
That's very fragile. Things will start breaking randomly as we serialize/deserialize various library types I'm afraid. The `FormatterServices.GetUninitializedObject()` via reflection path will not change the current behavior at all. So it should be 100% safe.

**Comment by dVakulen** [comment]
Sure. When I was implementing the `Activator.CreateInstance` it seemed the best way of doing it. Now, when we know that there's still exists `FormatterServices.GetUninitializedObject` - I can redo that implementation using this call..

**Comment by sergeybykov** [comment]
I think it's worth changing it for now, so that we can validate the rest between now and CoreCLR RTM. We will remove the workaround later when `FormatterServices.GetUninitializedObject()` is added back.

**Comment by jdom** [comment]
@dVakulen if you do change it, please also revert the additional default constructors that were added by that PR, which should no longer be needed.

**Comment by jdom** [comment]
BTW, I'll have a PR soon with that fix, so don't bother doing it

**Comment by jdom** [comment]
Reverted the change in #1745

**Comment by jdom** [comment]
Given that people on this thread might not be subscribed to the external github issue, I posted my findings on the exception serialization workaround here: https://github.com/dotnet/corefx/issues/7938#issuecomment-227580931

TL;DR: We are not blocked on serializing exceptions, by using a lot of reflection (which hopefully works in every platform, but I haven’t tested it), and with some strong limitations on what parts of the exception are deep copied/serialized. Still we will welcome proper support for built in serialization in a future version of .NET core, as it was announced.
